enhancements/oc/top-pvc-usage.md

```diff
@@ -0,0 +1,117 @@
+---
+title: top-pvc-usage
+authors:
+- "@gmeghnag"
+- "@tsmetana"
+reviewers:
+- "@dobsonj"
+- "@ardaguclu"
+approvers:
+- "@ardaguclu"
+creation-date: 2024-10-23
+last-updated: 2024-10-23
+status:
+see-also:
+replaces:
+superseded-by:
+---
+
+# top pvc
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Provide a simple command to display a PersistentVolumeClaim capacity usage.
+
+## Motivation
+
+There already exist options like `oc adm top pod` that can display usage
+statistic of the server resources. Adding an option to display also usage
+of the PersistentVolumeClaims seems like a natural extension which provides
+important information about the persistent storage available to the workloads.
+
+The volume usage statistics are already available via the web console but we
+should also have a way to get this through the CLI too.
+
+### Goals
+
+1. Provide a simple CLI option to display filesystem usage of
+   PersistentVolumeClaims.
+2. Display only the percentual usage for a given PersistentVolumeClaim or all
+   PersistentVolumeClaims in a given namespace.
+
+### Non-Goals
+
+## Proposal
+
+Implement `oc adm top persistentvolumeclaims` command that would show usage
+statistic of the bound PersistentVolumeClaim like this:
+
+```text
+oc adm top pvc -n reproducer-pvc
+NAMESPACE      NAME               USAGE(%) 
+reproducer-pvc pvc-reproducer-pvc 98.28    
+reproducer-pvc pvc-test-pvc       14.56   
+```
+
+### User Stories
+
+#### Story 1
+
+As an OCP project user, I want to see a list of all PVC's in my namespace and
+their space consumption so I know if my application has enough storage space.
+
+#### Story 2
+
+As an OCP cluster admin, I want to see a list of all PVC's on the cluster and
+their space consumption so I know if any components are running out of storage
+space.
+
+### Implementation Details/Notes/Constraints
+
+There are the `kubelet_volume_stats_used_bytes` and
```

Comment:
In a case where Prometheus pods are both down (which means that the `ClusterOperator/monitoring` is `Degraded`) the command would fail with the following error message:
```
$ oc adm top persistentvolumeclaims 
error: failed to get persistentvolumeclaims from Prometheus: unable to get /api/v1/query from URI in the openshift-monitoring/prometheus-k8s Route: prometheus-k8s-openshift-monitoring.apps.sharedocp415.lab.local->GET status code=503
```
In the same case the `oc adm top pod` and `oc adm top node` would fail as well:
```
$ oc adm top pod
error: Metrics not available for pod openshift-monitoring/alertmanager-main-0, age: 81h52m43.855409s
$ oc adm top nodes
error: metrics not available yet
```

Reply:
I think, we should raise  a better error message like `metrics not available yet` for this case. 

FWIW;
If I was the driver of this KEP, I'd first get the feedback from upstream about adding this first in there by extending https://github.com/kubernetes/kubernetes/blob/18f3941c24b9ed40eecd384f78778d2ff185c31d/staging/src/k8s.io/metrics/pkg/apis/metrics/v1beta1/types.go#L50 .  In my opinion, this is a useful feature and I believe upstream would consider this. Also if this is accepted by upstream, we'll have a natural client instead of sending a direct request to metrics server.

Reply:
- _"I think, we should raise a better error message like metrics not available yet for this case."_
  Ok, I can work on this.